peerdiversity/filter.go

```diff
@@ -0,0 +1,304 @@
+	// grouping info should be allowed/rejected by the Filter. This will be called ONLY
+	// AFTER the peer has successfully passed all of the Filter's internal checks.
+	// Note: If the peer is deemed accepted because of a whitelisting criteria configured on the Filter,
+	// the peer will be allowed by the Filter without calling this function.
+	// Similarly, if the peer is deemed rejected because of a blacklisting criteria
+	// configured on the Filter, the peer will be rejected without calling this function.
+	Allow(PeerGroupInfo) (allow bool)
+
+	// Increment is called by the Filter when a peer with the given Grouping Info.
+	// is added to the Filter state. This will happen after the peer has passed
+	// all of the Filter's internal checks and the Allow function defined above for all of it's Groups.
+	Increment(PeerGroupInfo)
+
+	// Decrement is called by the Filter when a peer with the given
+	// Grouping Info is removed from the Filter. This will happen when the caller/user of the Filter
+	// no longer wants the peer and the IP groups it belongs to to count towards the Filter state.
+	Decrement(PeerGroupInfo)
+
+	// PeerAddresses is called by the Filter to determine the addresses of the given peer
+	// it should use to determine the IP groups it belongs to.
+	PeerAddresses(peer.ID) []ma.Multiaddr
+}
+
+// Filter is a peer diversity filter that accepts or rejects peers based on the blacklisting/whitelisting
+// rules configured AND the diversity policies defined by the implementation of the PeerIPGroupFilter interface
+// passed to it.
+type Filter struct {
+	mu sync.Mutex
+	// An implementation of the `PeerIPGroupFilter` interface defined above.
+	pgm        PeerIPGroupFilter
+	peerGroups map[peer.ID][]PeerGroupInfo
+	// whitelisted Networks
+	wls cidranger.Ranger
+	// blacklisted Networks.
+	bls cidranger.Ranger
+	// legacy IPv4 Class A networks.
+	legacyCidrs cidranger.Ranger
+
+	logKey string
+
+	cplFnc func(peer.ID) int
+
+	cplPeerGroups map[int]map[peer.ID][]PeerIPGroupKey
+	// TODO This is for testing/logging purpose ONLY and can/should be removed later.
+	//cplRejections map[int]map[peer.ID]string
+}
+
+func NewFilter(pgm PeerIPGroupFilter, appName string, cplFnc func(peer.ID) int) (*Filter, error) {
+	if pgm == nil {
+		return nil, errors.New("peergroup implementation can not be nil")
+	}
+
+	// Crate a Trie for legacy Class N networks
+	legacyCidrs := cidranger.NewPCTrieRanger()
+	for _, cidr := range legacyClassA {
+		_, nn, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, err
+		}
+		if err := legacyCidrs.Insert(cidranger.NewBasicRangerEntry(*nn)); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Filter{
+		pgm:           pgm,
+		peerGroups:    make(map[peer.ID][]PeerGroupInfo),
+		wls:           cidranger.NewPCTrieRanger(),
+		bls:           cidranger.NewPCTrieRanger(),
+		legacyCidrs:   legacyCidrs,
+		logKey:        appName,
+		cplFnc:        cplFnc,
```

Comment:
The reason is to not have a dependency on the `kbucket` package.

I'll fix this if we move this to the DHT package.

Reply:
Ah, so trying to avoid circular dependencies? This will likely cause problems even in the DHT package (RT still needs Filter and Filter still needs RT). However, if the RT takes an interface this goes away.

Reply:
As discussed offline, let's punt this for now and we can pay this tech debt later if required.